### PR TITLE
Added labels in SubmitContainerTokenRequest

### PR DESCRIPTION
**Pull request by munishchouhan**
This PR will add labels in SubmitContainerTokenRequest class, so that it can be added in spack and conda docker file

### CODE REVIEW

**Comment by pditommaso** [comment]
Umm should it be a `List` of pairs or `Map` of key-value ?  a list would allow repetitions. Is this possible with docker labels?

**Comment by munishchouhan** [comment]
If there is a repeated label, but with a different value then the latest value will be applied, so i think we can use Map here
Screenshot 2024-03-12 at 17 16 16

**Comment by munishchouhan** [comment]
@pditommaso @marcodelapierre please review

**Comment by marcodelapierre** [comment]
Looks good to me, thanks @munishchouhan , will confirm after reviewing the other related PRs.

File: wave-api/src/main/java/io/seqera/wave/api/ContainerConfig.java [hunk 1]
```diff
@@ -50,12 +53,13 @@ public ContainerConfig() {
...
     }
 
-    public ContainerConfig(List<String> entrypoint, List<String> cmd, List<String> env, String workDir, List<ContainerLayer> layers) {
+    public ContainerConfig(List<String> entrypoint, List<String> cmd, List<String> env, String workDir, List<ContainerLayer> layers, Map<String, String> labels) {
         this.entrypoint = entrypoint;
         this.cmd = cmd;
         this.env = env;
         this.workingDir = workDir;
         this.layers = layers;
+        this.labels = labels;
```
Comment (marcodelapierre) [inline wave-api/src/main/java/io/seqera/wave/api/ContainerConfig.java:62]: shouldn't the the default (no args) initialiser above also define an empty `labels` object?
i.e. line 54
```
labels = new HashMap<>();
```

File: wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy [hunk 2]
```diff
@@ -51,50 +51,50 @@ class ContainerConfigTest extends Specification {
...
-        new ContainerConfig([], null, null, null, null).empty()
-        new ContainerConfig(null, [], null, null, null).empty()
-        new ContainerConfig(null, null, [], null, null).empty()
-        new ContainerConfig(null, null, null, '', null).empty()
-        new ContainerConfig(null, null, null, null, []).empty()
+        new ContainerConfig([], null, null, null, null, null).empty()
+        new ContainerConfig(null, [], null, null, null, null).empty()
+        new ContainerConfig(null, null, [], null, null, null).empty()
+        new ContainerConfig(null, null, null, '', null, null).empty()
+        new ContainerConfig(null, null, null, null, [], null).empty()
```
Comment (marcodelapierre) [inline wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy:68]: I would also test for the empty `labels` case, adding a line at the end, along the lines of:
```
new ContainerConfig(null, null, null, null, null, Map.of()).empty()
```

File: wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy [hunk 3]
```diff
@@ -51,50 +51,50 @@ class ContainerConfigTest extends Specification {
...
-        !new ContainerConfig(['x'], null, null, null, null).empty()
-        !new ContainerConfig(null, ['x'], null, null, null).empty()
-        !new ContainerConfig(null, null, ['x'], null, null).empty()
-        !new ContainerConfig(null, null, null, 'x', null).empty()
-        !new ContainerConfig(null, null, null, null, [new ContainerLayer()]).empty()
+        !new ContainerConfig(['x'], null, null, null, null, null).empty()
+        !new ContainerConfig(null, ['x'], null, null, null, null).empty()
+        !new ContainerConfig(null, null, ['x'], null, null, null).empty()
+        !new ContainerConfig(null, null, null, 'x', null, null).empty()
+        !new ContainerConfig(null, null, null, null, [new ContainerLayer()], [:]).empty()
```
Comment (marcodelapierre) [inline wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy:74]: I would break in 2 distinct lines, keeping the cases of `layers` and `labels` distinct, e.g.:
```
!new ContainerConfig(null, null, null, null, [new ContainerLayer()], null).empty()
!new ContainerConfig(null, null, null, null, null, Map.of('k', 'v')).empty()
```

**Review by marcodelapierre — CHANGES_REQUESTED** [review]
see minor requests - thank you Munish

File: wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy (hunk 2 above)
Comment (munishchouhan) [reply wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy:68]: done

File: wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy (hunk 3 above)
Comment (munishchouhan) [reply wave-api/src/test/groovy/io/seqera/wave/api/ContainerConfigTest.groovy:74]: done

File: wave-api/src/main/java/io/seqera/wave/api/ContainerConfig.java (hunk 1 above)
Comment (munishchouhan) [reply wave-api/src/main/java/io/seqera/wave/api/ContainerConfig.java:63]: done

**Review by marcodelapierre — APPROVED** [review]

**Comment by munishchouhan** [comment]
> @munishchouhan what do you think if we modelled `labels` as List instead of Map?
> 
> In the end, this would be the same approach adopted for the environment variables, i.e. `env` attribute; the two attributes look very similar to me. And then implementation of `labels` in Wave could closely mimick that of `env`.
> 
> What do you think?
> 
> See commit [618b846](https://github.com/seqeralabs/libseqera/commit/618b8465520e5c452b953346b362cebc1b4e51ae)

Map is used because the list will allow repetitions 
check this comment https://github.com/seqeralabs/libseqera/pull/18#issuecomment-1992052056

**Comment by marcodelapierre** [comment]
Thanks for pointing out Munish, had overlooked that comment.   
I see that Docker does work with multiple definitions of a same LABEL, by retaining the most recent one, a bit like environment variables definition in a Linux shell. As such, a List would have worked as well (same as it works for env variables).

This being said, I like your latest implementation, where the List is converted to Map only in the Wave server - it keeps the implementation of the functionality simple for API and clients.  I ultimately don't see big drawbacks of using the Map there, so I am leaving to you the ultimate decision.

@pditommaso this Libseqera PR is ready for merging.
